src/SeaportLeverage.sol

```diff
@@ -0,0 +1,267 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.24;
+
+import { SeaportBase } from "./SeaportBase.sol";
+
+import { IIonPool } from "./interfaces/IIonPool.sol";
+import { IGemJoin } from "./interfaces/IGemJoin.sol";
+import { IWhitelist } from "./interfaces/IWhitelist.sol";
+import {
+    Order,
+    OrderParameters,
+    OfferItem,
+    ConsiderationItem,
+    ItemType
+} from "seaport-types/src/lib/ConsiderationStructs.sol";
+
+import { IERC20 } from "openzeppelin-contracts/contracts/interfaces/IERC20.sol";
+import { SafeERC20 } from "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { WadRayMath } from "ion-protocol/src/libraries/math/WadRayMath.sol";
+
+contract SeaportLeverage is SeaportBase {
+    using SafeERC20 for IERC20;
+    using WadRayMath for uint256;
+
+    uint256 internal constant TSLOT_RESULTING_ADDITIONAL_COLLATERAL = 2;
+
+    IWhitelist public immutable WHITELIST;
+
+    // Offer item validation, the collateral asset is being offered
+    error OTokenMustBeCollateral(address token);
+    error OStartMustBeCollateralToPurchase(uint256 startAmount, uint256 collateralToPurchase);
+    error OEndMustBeCollateralToPurchase(uint256 endAmount, uint256 collateralToPurchase);
+
+    // Consideration item 1 validation, only used as a callback
+    error C1StartAmountMustBeAmountToBorrow(uint256 startAmount, uint256 amountToBorrow);
+    error C1EndAmountMustBeAmountToBorrow(uint256 endAmount, uint256 amountToBorrow);
+
+    // Consideration item 2 validation, the base asset is beign paid for the offer
+    error C2TokenMustBeBase(address token);
+    error C2StartMustBeAmountToBorrow(uint256 startAmount, uint256 amountToBorrow);
+    error C2EndMustBeAmountToBorrow(uint256 endAmount, uint256 amountToBorrow);
+
+    /**
+     * @notice Only allows whitelisted borrowers to use this contract.
+     * @dev This contract will be a part of `protocolWhitelist`, so all
+     * calls made from this contract to the IonPool as the sender will
+     * succeed. This contract verifies the whitelist proof on its own
+     * in order to avoid having to pass the proof through the callbacks
+     * and to the IonPool.
+     * @param proof Merkle path for the whitelist root.
+     */
+    modifier onlyWhitelistedBorrowers(bytes32[] memory proof) {
+        WHITELIST.isWhitelistedBorrower(ILK_INDEX, msg.sender, msg.sender, proof);
+        _;
+    }
+
+    constructor(
+        IIonPool pool,
+        IGemJoin gemJoin,
+        uint8 ilkIndex,
+        IWhitelist whitelist
+    )
+        SeaportBase(pool, gemJoin, ilkIndex)
+    {
+        WHITELIST = whitelist;
+
+        // The IonPool takes the initial deposit collateral from
+        // the caller and seaport takes base asset from the offerer.
+        BASE.approve(address(SEAPORT), type(uint256).max);
+        COLLATERAL.approve(address(POOL), type(uint256).max);
```

Comment:
This is shouldn't be necessary since the POOL doesn't actually hold any of the collateral. Let's fix the comment above as well since it is the `GemJoin` that takes the collateral.